Mathlib/Data/Nat/FinMulAntidiagonal.lean

```diff
@@ -0,0 +1,251 @@
+/-
+Copyright (c) 2024 Arend Mellendijk. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Arend Mellendijk
+-/
+
+import Mathlib.Algebra.Order.Antidiag.Pi
+import Mathlib.NumberTheory.ArithmeticFunction
+
+/-!
+# Sets of tuples with a fixed product
+
+This file defines the finite set of `d`-tuples of natural numbers with a fixed product `n` as
+`Nat.finMulAntidiagonal`.
+
+## Main Results
+* There are `d^(ω n)` ways to write `n` as a product of `d` natural numbers, when `n` is squarefree
+(`card_finMulAntidiagonal_of_squarefree`)
+-/
+
+open Finset
+open scoped BigOperators ArithmeticFunction
+namespace PNat
+
+/-- The set of divisors of a positive natural number.
+This is `Nat.divisorsAntidiagonal` without a special case for `n = 0`. -/
+def divisorsAntidiagonal (n : ℕ+) : Finset (ℕ+ × ℕ+) :=
+  (Nat.divisorsAntidiagonal n).attach.map
+    ⟨fun x =>
+      (⟨x.val.1, Nat.pos_of_mem_divisors <| Nat.fst_mem_divisors_of_mem_antidiagonal x.prop⟩,
+      ⟨x.val.2, Nat.pos_of_mem_divisors <| Nat.snd_mem_divisors_of_mem_antidiagonal x.prop⟩),
+    fun _ _ h => Subtype.ext <| Prod.ext (congr_arg (·.1.val) h) (congr_arg (·.2.val) h)⟩
+
+@[simp]
+theorem mem_divisorsAntidiagonal {n : ℕ+} (x : ℕ+ × ℕ+) :
+    x ∈ divisorsAntidiagonal n ↔ x.1 * x.2 = n := by
+  simp_rw [divisorsAntidiagonal, Finset.mem_map, Finset.mem_attach, Function.Embedding.coeFn_mk,
+    Prod.ext_iff, true_and, ← coe_inj, Subtype.exists]
+  aesop
+
+instance instHasAntidiagonal : Finset.HasAntidiagonal (Additive ℕ+) where
+  antidiagonal n := divisorsAntidiagonal (Additive.toMul n) |>.map
+    (.prodMap (Additive.ofMul.toEmbedding) (Additive.ofMul.toEmbedding))
+  mem_antidiagonal {n a} := by
+    obtain ⟨a₁, a₂⟩ := a
+    revert n a₁ a₂
+    simp_rw [Additive.ofMul.surjective.forall, ← ofMul_mul, Additive.ofMul.injective.eq_iff]
+    simp
+
+@[norm_cast]
+theorem coe_prod {ι : Type*} (f : ι → ℕ+) (s : Finset ι) :
+    ↑(∏ i in s, f i) = (∏ i in s, f i : ℕ) :=
```

Comment:
```suggestion
    ↑(∏ i ∈ s, f i) = (∏ i ∈ s, f i : ℕ) :=
```
can you move that earlier?

Reply:
Sadly none of the PNat files import `Finset.prod`. I've moved it to `Mathlib.Algebra.Order.BigOperators.Ring.Finset` next to a similar lemma about `abs`. 

If anybody knows a better place please tell me and I'll move the lemma. `#find_home` says `[Mathlib.RingTheory.Coprime.Lemmas, Mathlib.Algebra.Order.BigOperators.Ring.Finset, Mathlib.Data.Nat.Choose.Sum, Mathlib.Tactic.Module, Mathlib.RingTheory.Ideal.Lattice]`

Reply:
Looks fine to me!